Value type null-restricted array support

theresa-m opened the pull request:
Please note this pull request adds support to create a null-restricted array but not a flattened array. That will come in a later change.

- create a null-restricted array field for every RAM class if one doesn't exist when a null-restricted array is created (see https://github.com/eclipse-openj9/openj9/issues/19460#issue-2284036228)
- implement JVM_IsNullRestrictedArray and JVM_NewNullRestrictedArray
- Update exception thrown when null is set to a null-restricted array from NullPointerException to ArrayStoreException as described in https://openjdk.org/jeps/8316779

Related: https://github.com/eclipse-openj9/openj9/issues/17340
Fixes: https://github.com/eclipse-openj9/openj9/issues/19460

After these changes:
- ValueTypeArrayTests have been enabled and pass https://github.com/eclipse-openj9/openj9/issues/19460#issuecomment-2163847738
- ValueTypeTestsJIT still has some failures https://github.com/eclipse-openj9/openj9/issues/19708
- ValueTypeArrayTestsJIT still has some failures https://github.com/eclipse-openj9/openj9/issues/19913
- ValueTypeSystemArraycopyTests pass for Xint variations https://github.com/eclipse-openj9/openj9/issues/19914

hangshao0 commented on runtime/vm/createramclass.cpp:
```diff
@@ -197,6 +197,22 @@ static void initializeClassLinks(J9Class *ramClass, J9Class *superclass, J9Memor
 
+static J9Class *
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+createdArrayClass(J9Class *elementClass, UDATA valueTypeFlags) {
```
`getArrayClass` is a better name.

This can be turned into a macro that takes a J9Class and a UDATA valueTypeFlags. We can always pass in 0 as valueTypeFlags in non-VT build.

ThanHenderson replied:
@hangshao0 out of curiosity, what is your reasoning behind preferring a macro to something like the following?

```
static inline J9Class *
getArrayClass(J9Class *elementClass, UDATA valueTypeFlags = 0)
{
#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
        if (J9_ARE_ALL_BITS_SET(valueTypeFlags, J9ClassArrayIsNullRestricted)) {
                return elementClass->nullRestrictedArrayClass;
        }
#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
        return elementCalss->arrayClass;
}
```

hangshao0 replied:
Macro is usually cheaper that a function call.

ThanHenderson replied:
True, but if it is specified as an `inline` function, given it's body, it would likely be inlined.

hangshao0 commented:
Before continue reviewing this, I am wondering if it is easier to add an new function
`internalCreateArrayClassWithOptions(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Class* elementClass, UDATA options)` (The existing `internalCreateArrayClass()` becomes a special case that passed in options is 0) and call this function in `JVM_NewNullRestrictedArray` with a new option flag `J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY` indicating it is creating a null restricted array.

ThanHenderson commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -763,15 +763,63 @@ JVM_IsImplicitlyConstructibleClass(JNIEnv *env, jclass cls)
+
+	if (NULL == newArray) {
+		vmFuncs->setHeapOutOfMemoryError(currentThread);
+		return NULL;
```
In this path, there is no call to `vmFuncs->internalExitVMToJNI(currentThread);`

If we are already using `goto`s here and `arrayRef` at this point is already `NULL`, we can just:
```
if (NULL == newArray) {
	vmFuncs->setHeapOutOfMemoryError(currentThread);
	goto done;
}

```

theresa-m replied:
Good catch, thanks.

ThanHenderson commented on runtime/vm/createramclass.cpp:
```diff
@@ -2093,8 +2109,21 @@ checkFlattenableFieldValueClasses(J9VMThread *currentThread, J9ClassLoader *clas
+			J9CLASS_IS_ARRAY(state->ramClass) &&
+			J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(((J9ArrayClass*)state->ramClass)->componentType) &&
+			J9_ARE_NO_BITS_SET(state->valueTypeFlags, J9ClassArrayIsNullRestricted)
+		))
```
The `&&` should be at the start of the line, and the enclosing `)` should be inline with the `if`.
```
if (FALSE == ((NULL != state->ramClass)
	        && J9CLASS_IS_ARRAY(state->ramClass)
	        && J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(((J9ArrayClass*)state->ramClass)->componentType)
	        && J9_ARE_NO_BITS_SET(state->valueTypeFlags, J9ClassArrayIsNullRestricted))
)
```

ThanHenderson commented on runtime/vm/createramclass.cpp:
```diff
@@ -2128,7 +2157,18 @@ internalCreateRAMClassDone(J9VMThread *vmThread, J9ClassLoader *classLoader, J9C
+		if (FALSE == (J9ROMCLASS_IS_ARRAY(romClass) &&
+				J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(elementClass) &&
+				J9_ARE_NO_BITS_SET(state->valueTypeFlags, J9ClassArrayIsNullRestricted)
+			))
```
Same thing with the `&&` here.

ThanHenderson commented on runtime/vm/createramclass.cpp:
```diff
@@ -3664,9 +3730,24 @@ internalCreateRAMClassFromROMClass(J9VMThread *vmThread, J9ClassLoader *classLoa
 		}
+	}
+
+	if (flattenedClassCache != (J9FlattenedClassCache *) flattenedClassCacheBuffer) {
```
Unnecessary space before `flattenedClassCacheBuffer`.

ThanHenderson commented on runtime/vm/createramclass.cpp:
```diff
@@ -2844,7 +2899,11 @@ internalCreateRAMClassFromROMClassImpl(J9VMThread *vmThread, J9ClassLoader *clas
+			ramClass = createdArrayClass(elementClass,
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+				*valueTypeFlags
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
```
The `,` needs to be on the same line as `*valueTypeFlags`, not outside of the `if` directive.

ThanHenderson commented on runtime/vm/createramclass.cpp:
```diff
@@ -2175,7 +2215,11 @@ internalCreateRAMClassDone(J9VMThread *vmThread, J9ClassLoader *classLoader, J9C
-				alreadyLoadedClass = elementClass->arrayClass;
+				alreadyLoadedClass = createdArrayClass(elementClass
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+				, state->valueTypeFlags
```
The syntax alignment of `, state->valueTypeFlags` seems a bit off, maybe add another tab or two.

ThanHenderson commented on runtime/vm/createramclass.cpp:
```diff
@@ -2374,7 +2418,11 @@ internalCreateRAMClassDone(J9VMThread *vmThread, J9ClassLoader *classLoader, J9C
+						alreadyLoadedClass = createdArrayClass(elementClass
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+						, state->valueTypeFlags
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
```
Same comment about the alignment.

ThanHenderson commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -3396,6 +3397,9 @@ typedef struct J9Class {
 	J9MemberNameListNode *memberNames;
 #endif /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+	struct J9Class* nullRestrictedArrayClass;
```
The `*` should be right aligned to match the other members, same comment for `companionArray`.

hangshao0 commented:
> Before continue reviewing this, I am wondering if it is easier to add an new function
internalCreateArrayClassWithOptions(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Class* elementClass, UDATA options) (The existing internalCreateArrayClass() becomes a special case that passed in options is 0) and call this function in JVM_NewNullRestrictedArray with a new option flag J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY indicating it is creating a null restricted array.

The reason for this is to see if we can minimize the change in `createramclass.cpp` and do not need special handling of pushing/poping the protectionDomain and class loading stack. The nullable J9arrayClass and null restricted J9arrayClass can be created separately.

theresa-m commented:
> > Before continue reviewing this, I am wondering if it is easier to add an new function
> > internalCreateArrayClassWithOptions(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Class* elementClass, UDATA options) (The existing internalCreateArrayClass() becomes a special case that passed in options is 0) and call this function in JVM_NewNullRestrictedArray with a new option flag J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY indicating it is creating a null restricted array.
> 
> The reason for this is to see if we can minimize the change in `createramclass.cpp` and do not need special handling of pushing/poping the protectionDomain and class loading stack. The nullable J9arrayClass and null restricted J9arrayClass can be created separately, one in each `internalCreateArrayClass*()` call.

I agree this approach may be cleaner, I'll give it a try and look over the review comments that are still applicable.

ThanHenderson commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -763,15 +763,62 @@ JVM_IsImplicitlyConstructibleClass(JNIEnv *env, jclass cls)
+	}
+
+	if (NULL == ramClass->nullRestrictedArrayClass) {
+		J9ROMArrayClass* arrayOfObjectsROMClass = (J9ROMArrayClass*)J9ROMIMAGEHEADER_FIRSTCLASS(vm->arrayROMClasses);
```
`J9ROMArrayClass* ...` -> `J9ROMArrayClass *...`

ThanHenderson commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -763,15 +763,62 @@ JVM_IsImplicitlyConstructibleClass(JNIEnv *env, jclass cls)
+
+	if (NULL == ramClass->nullRestrictedArrayClass) {
+		J9ROMArrayClass* arrayOfObjectsROMClass = (J9ROMArrayClass*)J9ROMIMAGEHEADER_FIRSTCLASS(vm->arrayROMClasses);
+		vmFuncs->internalCreateArrayClassWithOptions(currentThread, arrayOfObjectsROMClass, ramClass,
```
https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#function-parameter-list-does-not-fit-on-one-line

Formatting for functions with params that exceed one line should be formatted like one of the two options in the document. However, note that even though that the examples in that document place the commas at the end of the lines, there is push back if they are not at the beginning of the next line.

theresa-m replied:
so the comma for ramClass would be like this?
```
		vmFuncs->internalCreateArrayClassWithOptions(
			currentThread, arrayOfObjectsROMClass, ramClass
			, J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY);
```

ThanHenderson replied:
I think the `J` should be on the same column as the above `c`.

ThanHenderson replied:
That is just what I have been told though. I personally think it is better with the comma on the preceding line.

ThanHenderson commented on runtime/vm/createramclass.cpp:
```diff
@@ -197,6 +197,18 @@ static void initializeClassLinks(J9Class *ramClass, J9Class *superclass, J9Memor
+		return elementClass->arrayClass;
+	}
+}
+
```
This would be better if it was reformatted like:
```
static inline J9Class *
getArrayClass(J9Class *elementClass, UDATA options)
{
#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
        if (J9_ARE_ALL_BITS_SET(options, J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY)) {
                return elementClass->nullRestrictedArrayClass;
        }
#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
        return elementCalss->arrayClass;
}
```

theresa-m replied:
I think the `else` makes it more readable personally.

ThanHenderson replied:
The `else` is just redundant if the body of the `if` has a `return`, and leads to funky formatting with the if-directives. Just my preference.

ThanHenderson reviewed (commented):
Just a few formatting comments.

theresa-m commented:
The JIT updates should be covered by https://github.com/eclipse-openj9/openj9/issues/19913.

I'll take a look at the GC code.

hangshao0 commented on runtime/gc_structs/ClassArrayClassSlotIterator.cpp:
```diff
@@ -46,10 +46,27 @@ GC_ClassArrayClassSlotIterator::nextSlot()
 			classPtr = (J9Class *)_iterateClazz->arrayClass;
 			if (!_isArrayClass) {
 				_state = classArrayClassSlotIterator_state_done;
+				break;
```
Why `break` here ?

dmitripivkine replied:
Right. `break` should not be here.

hangshao0 commented on runtime/gc_structs/ClassArrayClassSlotIterator.cpp:
```diff
@@ -46,10 +46,27 @@ GC_ClassArrayClassSlotIterator::nextSlot()
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if (NULL == classPtr) {
+				classPtr = (J9Class *)_iterateClazz->nullRestrictedArrayClass;
+				_state += classArrayClassSlotIterator_state_componentType;
```
`_state` should not be incremented by `classArrayClassSlotIterator_state_componentType`.

hangshao0 replied:
It is possible that `classPtr` (`_iterateClazz->nullRestrictedArrayClass`) is null and it is returned incorrectly from here.

dmitripivkine replied:
This iterator is written in style, where if class pointer to return is NULL we are going to the next state. Do we need to restrict to report `nullRestrictedArrayClass` otherwise `companionArray`. Can (and should) we read both unconditionally and probably rely on one or another is NULL? If both of this fields can be non-NULL should we report both to have proper class liveness discovery?

dmitripivkine replied:
If I understand intention correctly, I think the only change in this method we should add two extra states to return class pointers from `nullRestrictedArrayClass` and `companionArray` fields, no extra logic required. State `classArrayClassSlotIterator_state_nullRestrictedArrayClass` and handling for it is proposed already.

hangshao0 commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -111,6 +111,18 @@ GC_ClassLoaderClassesIterator::nextClass()
 				/* this class is defined in the loader, so follow its array classes */
 				_nextClass = result->arrayClass;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if ( (result->classLoader == _classLoader) && (NULL != result->nullRestrictedArrayClass) ) {
```
It seems that the logic here could leak some array classes. In the original code, if ClassA has 2D arrays, `nextClass()` returns classes in the following sequence: ClassA, 1D-arrayClassA, 2D-ArrayClassA, ClassB, etc.

However if ClassA is a flattenable VT, `nextClass()` in the new code returns classes as: ClassA, 1D-arrayClassA, 2D-arrayClassA, ClassB, etc. When `result` is 2D-arrayClassA, both `else if` checks added here will be false (nullrestrictedArray can only be 1-D). 1D-nullRestricedArrayClassA is leaked.

hangshao0 commented:
@dmitripivkine I added you to review the GC change here. 

The change here is to support value type which could have 2 types of arrays. One allows its element to be null, one does not allow its element to be null (null restricted array). A new field `nullRestrictedArrayClass` is added in `J9Class`. (Note `nullRestrictedArrayClass` can only be 1D array).

dmitripivkine commented on runtime/gc_structs/ClassArrayClassSlotIterator.cpp:
```diff
@@ -46,10 +46,27 @@ GC_ClassArrayClassSlotIterator::nextSlot()
+				if (NULL == ((J9ArrayClass *)classPtr)->companionArray) {
+					_state += 1; /* skip nullRestrictedArrayClass case */
+				}
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
```
This block is in a wrong place. This check should be done inside `classArrayClassSlotIterator_state_nullRestrictedArrayClass` state.

dmitripivkine commented:
With changes in GC iterators correspondent DDR code should be updated as well (like `GCClassArrayClassSlotIterator.java`)

theresa-m commented:
> With changes in GC iterators correspondent DDR code should be updated as well (like `GCClassArrayClassSlotIterator.java`)

Is there one for ClassLoaderClassesIterator.cpp? I'm not seeing any.

dmitripivkine commented:
> > With changes in GC iterators correspondent DDR code should be updated as well (like `GCClassArrayClassSlotIterator.java`)
> 
> Is there one for ClassLoaderClassesIterator.cpp? I'm not seeing any.

Yes, there is no designated iterator. Currently I am trying to make my IDE to work after update. I will double check.

dmitripivkine commented on runtime/gc_structs/ClassArrayClassSlotIterator.cpp:
```diff
@@ -50,6 +50,12 @@ GC_ClassArrayClassSlotIterator::nextSlot()
+			classPtr = ((J9ArrayClass *)_iterateClazz)->companionArray;
+			_state += 1;
+			break;
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
```
Should we have another state to return `nullRestrictedArrayClass` as well?

theresa-m replied:
Yes, this is missing the case where a nullRestrictedArrayClass exists and an arrayClass doesn't.

theresa-m replied:
I updated this file.

dmitripivkine replied:
Please explain why `nullRestrictedArrayClass` should not be returned? Currently iterator returns all slots of `J9ArrayClass` can contain pointer to another class regardless any application logic.

theresa-m replied:
Okay. I will add a case for nullRestrictedArrayClass then since it may contain a pointer.

theresa-m replied:
Updated.

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -108,9 +108,22 @@ GC_ClassLoaderClassesIterator::nextClass()
 				_nextClass = result->arrayClass;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if ((result->classLoader == _classLoader)
+				&& J9_IS_J9ARRAYCLASS_NULL_RESTRICTED(result)
```
Do we need this condition? Current assumption is any slot can contain pointer to j9class should be scanned unconditionally except it is NULL.

theresa-m replied:
Yes, otherwise there could be an infinite loop. If there exists one array class of arity one and a nullRestrictedArrayClass object for a base ram class they will be pointing to each other as companion classes. 
```
   baseClass-->arrayClass
           |
           | --->nullRestrictedArrayClass
```
arrayClass->companionClass is nullRestrictedArraClass
nullRestrictedArrayClass->companionClass is arrayClass

hangshao0 replied:
Please note that null restricted array can only be 1D array, but you can still have a 2D nullable array whose element is 1D null restricted array. So it is possible to have things like `A, A[], A[][], A[][][], A[!], A[][!], A[][][!]` (`[!]` means null is not allowed in that dimension and` [!]` can only be at the last dimension):
```

       |------->arrayClass-->arrayClass-->arrayClass
       |          A[]           A[][]       A[][]
       |                   
baseClass (A)     
       |                   
       |           A[!]                   A[][!]      A[][][!]
       |----> nullRestrictedArrayClass-->ArrayClass-->ArrayClass
```
Only `A[]` and` A[!]` points to each other as companions.

You need sth like:

```
J9Class* getNullRestrictedArray(J9Class* c) {
     J9Class* baseClass = NULL;
     if (NULL == c->leafComponentType) {
        /* c is a baseClass*/ 
        baseClass = c;
     } else {
         /* c is an arrayClass,  c->leafComponentType is the baseClass*/ 
        baseClass = c->leafComponentType;
     }
     return baseClass->nullRestrictedArraClass;
}

if ((result->classLoader == _classLoader) && (NULL != result->arrayClass)) {
	/* this class is defined in the loader, so follow its array classes */
	_nextClass = result->arrayClass;
else if ((result->classLoader == _classLoader) && (NULL != getNullRestrictedArray(result))) {
	_nextClass = getNullRestrictedArray(result);
} else if (TABLE_CLASSES == _mode) {
	_nextClass = nextTableClass();
} else {
	_nextClass = nextSystemClass();
}
```

theresa-m replied:
That looks good to me. I added your suggestion to the change.

dmitripivkine commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCClassArrayClassSlotIterator.java:
```diff
@@ -47,6 +47,11 @@ protected GCClassArrayClassSlotIterator(J9ClassPointer clazz) throws CorruptData
+		if(slot.notNull()) {
+			slots.add(slot);
+			addresses.add(VoidPointer.cast(clazz.nullRestrictedArrayClassEA()));
+		}
```
Should we add `companionArray` as well?

theresa-m replied:
class.nullRestrictedArrayClass is the same object as class.arrayClass.companionArray

dmitripivkine commented on runtime/gc_structs/ClassArrayClassSlotIterator.cpp:
```diff
@@ -42,6 +42,16 @@ GC_ClassArrayClassSlotIterator::nextSlot()
+				_state += 1;
+			}
+			break;
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
```
Why you put this case first? From GC point of view does not matter how `nullRestrictedArrayClass` and `arrayClass` are relate, it just another field.
`if (!_isArrayClass)` should be located in the first case only (Iterator requested to walk non-array class.
So, I think you can remove `_isArrayClass` check, just keep state increment and move this case lower together with `classArrayClassSlotIterator_state_companionArray` case.

theresa-m replied:
Okay. So then I think the nullRestrictedArrayClass case isn't needed at all. The nullRestrictedArrayClass field will only contain an object for a ram class that is a base array. The same object would be accessible as the companionArray field of an array ram class.

theresa-m replied:
It is possible that a nullRestrictedArrayClass field exists when an arrayClass does not exist. Without the nullRestrictedArrayClass case this would be missed.

theresa-m replied:
Okay. The nullRestrictedArrayClass case will not be needed then since _iterateClazz is expected to be an array

hangshao0 commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -108,9 +123,13 @@ GC_ClassLoaderClassesIterator::nextClass()
 				/* this class is defined in the loader, so follow its array classes */
 				_nextClass = result->arrayClass;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if ((result->classLoader == _classLoader) && getNullRestrictedArray(result)) {
```
`getNullRestrictedArray(result)` here should be `(NULL != getNullRestrictedArray(result))`

theresa-m replied:
This is fixed.

hangshao0 commented:
Do you have more review comments ? @dmitripivkine

tajila commented on runtime/vm/createramclass.cpp:
```diff
@@ -197,6 +197,18 @@ static void initializeClassLinks(J9Class *ramClass, J9Class *superclass, J9Memor
 #endif /* JAVA_SPEC_VERSION == 8 */
 
+static inline J9Class *
+getArrayClass(J9Class *elementClass, UDATA options) {
```
use `VMINLINE` instead

tajila commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -763,15 +763,62 @@ JVM_IsImplicitlyConstructibleClass(JNIEnv *env, jclass cls)
+	if (NULL == obj) {
+		vmFuncs->setCurrentException(currentThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+	} else {
+		jclass clazz = env->GetObjectClass(obj);
```
avoid using JNI functions (anything with `env->*`) once you've acquired VMAccess as this will internally acquire and release VMaccess. Instead you can simply unwrap the `obj` with `J9_JNI_UNWRAP_REFERENCE` and then get the object class with `J9OBJECT_CLAZZ`

tajila commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -763,15 +763,62 @@ JVM_IsImplicitlyConstructibleClass(JNIEnv *env, jclass cls)
 JNIEXPORT jboolean JNICALL
 JVM_IsNullRestrictedArray(JNIEnv *env, jobject obj)
```
it doesn't need to be done in this PR, but all Valhalla related functions should be moved to a different file (perhaps valhallavmi.cpp). For large features like valhalla, loom, etc. we should try to avoid bundling  in the same file like we did in the past.

theresa-m replied:
I'll do this in a separate pr.

tajila commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -763,15 +763,62 @@ JVM_IsImplicitlyConstructibleClass(JNIEnv *env, jclass cls)
+			goto done;
+		}
+	}
+
```
after classloading the j9class could be redefined, so you need to fetch the current one with `VM_VMHelpers::currentClass`

tajila commented on runtime/gc_structs/ClassArrayClassSlotIterator.cpp:
```diff
@@ -50,6 +50,16 @@ GC_ClassArrayClassSlotIterator::nextSlot()
 			}
 			break;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+		case classArrayClassSlotIterator_state_nullRestrictedArrayClass:
```
IIRC should this state do the following

```
		if (!_isArrayClass) {
				_state = classArrayClassSlotIterator_state_done;
			} else {
```

Given that a non-array class now potentially points to two array classes?

dmitripivkine replied:
Well, I think not here at least. This iterator is one of collection of iterators that `GC_ClassIteratorClassSlots` calls. This particular one suppose to return extra specific slots if given class is an array class. So, this check means "skip is given class is not an array class". If class is not an array class this iterator should return nothing in my opinion. However there is a question should any other iterator from this collection return new slots?

theresa-m replied:
If _IsArrayClass is true there will never be a nullRestrictedArrayClass, so maybe this case isn't needed at all.

tajila replied:
Okay, I understand. Its fine as is.

tajila replied:
My previous comment was in response to Dmitri

> If _IsArrayClass is true there will never be a nullRestrictedArrayClass, so maybe this case isn't needed at all.

Yes, since `companionArray` and `nullRestrictedArrayClass` are the same slot then we only need one which can be `companionArray`

theresa-m replied:
I removed the nullRestrictedArrayClass case.

tajila commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -763,15 +763,62 @@ JVM_IsImplicitlyConstructibleClass(JNIEnv *env, jclass cls)
+		goto done;
+	}
+
+	arrayRef = (jarray)vmFuncs->j9jni_createLocalRef(env, newArray);
```
this can also fail, it would be a native OOM

tajila replied:
NVM, since you are only creating one in this function you are guaranteed a slot

tajila commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -763,15 +763,62 @@ JVM_IsImplicitlyConstructibleClass(JNIEnv *env, jclass cls)
+
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+	ramClass = J9VMJAVALANGCLASS_VMREF(currentThread, J9_JNI_UNWRAP_REFERENCE(componentType));
+
```
you should also check negative array size and throw `J9VMCONSTANTPOOL_JAVALANGNEGATIVEARRAYSIZEEXCEPTION`

tajila commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -108,9 +123,13 @@ GC_ClassLoaderClassesIterator::nextClass()
 				/* this class is defined in the loader, so follow its array classes */
 				_nextClass = result->arrayClass;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if ((result->classLoader == _classLoader) && (NULL != getNullRestrictedArray(result))) {
```
Spoke with Hang. Given that the higher order J9Class arrays are different (ie. Foo[][] != Foo[][!], etc.) then for a given J9Class we need to potentially iterate two lists. The normal array list and the NR list

One approach is to add a new state `VALUE_CLASSES` for the Null Restricted case where, clazz->NRArray is walked (and then the subsequent arrayClazz->arrayClazz) until NULL is reached. At this point, we will return to the base class via leafComponentType then switch back to `TABLE_CLASSES` mode. In this mode, it should work like it currently does.

Here is an example:

```
	if (NULL != result) {
		if (ANONYMOUS_CLASSES == _mode) {
			_nextClass = nextAnonymousClass();
		} else {
			if ((result->classLoader == _classLoader) && (NULL != result->nullRestrictedArrayClass)) {
				_mode = VALUE;
				_nextClass = result->nullRestrictedArrayClass
			} else if ( (result->classLoader == _classLoader) && (NULL != result->arrayClass) ) {
				/* this class is defined in the loader, so follow its array classes */
				_nextClass = result->arrayClass;
			} else if (VALUE == _mode) {
				J9ArrayClass *arrayClazz = result->leafComponentType->arrayClass;
				_mode = TABLE_CLASSES;
				if ( (result->classLoader == _classLoader) && (NULL != arrayClazz)) {
					_nextClass = arrayClazz;
				} else {
					_nextClass = nextTableClass();
				}
			} else if (TABLE_CLASSES == _mode) {
				_nextClass = nextTableClass();
			} else {
				_nextClass = nextSystemClass();
			}
		}
	}
```

Another option is we create an `ARRAY_MODE` and encapsulate all array walking logic inside a `nextArrayClass()` function. This would clean up the code above, but we would still require a `VALUE` mode or an equivalent substate to walk the two lists.

@dmitripivkine do you have a preference on the suggested approach?

dmitripivkine replied:
I am not sure use `_mode` is a good idea - you need to preserve original value (ex. `TABLE_CLASSES`) and restore it after. I think (maybe) creating an independent `_state` will serve better. For example `_mode` is going to stay unchanged (and there is no assumption only `TABLE_CLASSES` might have Null Restricted). You can switch between "Array" and "VT" states and still be in the current mode. Will it work?

tajila replied:
>  creating an independent _state will serve better.

Yes this will work.

theresa-m replied:
I updated my changes to include this. Please take a look.

hangshao0 replied:
> I updated my changes to include this. Please take a look.

I think what Dmitri means by _state are enum like https://github.com/eclipse-openj9/openj9/blob/1cac5050a7914a28fbab9b1e8e1aea9c16eeaa5a/runtime/gc_structs/ClassArrayClassSlotIterator.hpp#L47-L52

Use "Array" state when walking the normal arrayclass list, use "VT" or "NullrestirctiedArray" state when walking the nullrestricted array list.

theresa-m replied:
I figured since there's only two states a boolean would be sufficient to describe that.

tajila commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -108,9 +108,25 @@ GC_ClassLoaderClassesIterator::nextClass()
+				if ( (result->classLoader == _classLoader) && (NULL != arrayClazz)) {
+					_nextClass = arrayClazz;
+				} else {
+					_nextClass = nextTableClass();
```
The only issue left is that the code assumes if `_nullRestrictedArrayWalkState` is true then `_mode == TABLE_CLASSES`. I think thats tru right now, but it may not be tru in the future.

dmitripivkine replied:
Skipping functional comments for now, style wise this is cpp code, not c and _nullRestrictedArrayWalkState is defined as bool. So, you should use true instead of TRUE, do if (_nullRestrictedArrayWalkState) instead of if (TRUE == _nullRestrictedArrayWalkState) etc. Also please add an initialization of the new variable in constructor.

hangshao0 commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -108,13 +152,18 @@ GC_ClassLoaderClassesIterator::nextClass()
 			_nextClass = nextAnonymousClass();
 		} else {
-			if ( (result->classLoader == _classLoader) && (NULL != result->arrayClass) ) {
+			J9Class* array = nextArrayClass();
```
This should be `J9ArrayClass*` and `nextArrayClass()` should return `J9ArrayClass* `

theresa-m replied:
My preference is to have less casting, I don't think it matters.

theresa-m commented:
@tajila and @dmitripivkine can you take a look at the latest version of ClassLoaderClassesIterator.cpp?

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.hpp:
```diff
@@ -62,6 +62,15 @@ class GC_ClassLoaderClassesIterator
+		classLoaderClassIterator_state_valueTypeArrayList,
+		classLoaderClassIterator_state_arrayList,
+		classLoaderClassIterator_state_done
+	};
```
You can use named enum and use this name to define variable:

    enum ArrayClassState {
        classLoaderClassIterator_state_valueTypeArray = 0,
        classLoaderClassIterator_state_valueTypeArrayList,
        classLoaderClassIterator_state_arrayList,
        classLoaderClassIterator_state_done
    };
    ArrayClassState _arrayState;

Also all class members should be initialized in Constructor.

dmitripivkine replied:
BTW state names are private, you can skip `classLoaderClassIterator` prefix if you like

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -108,13 +152,18 @@ GC_ClassLoaderClassesIterator::nextClass()
 			} else if (TABLE_CLASSES == _mode) {
 				_nextClass = nextTableClass();
+				_startingClass = _nextClass;
+				_arrayState = 0;
```
classLoaderClassIterator_state_valueTypeArray should be used instead of 0

dmitripivkine replied:
BTW these states are private, so can use names without `classLoaderClassIterator` prefix if you like

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -99,6 +101,48 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+					_iterateArrayClazz = _startingClass->nullRestrictedArrayClass;
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
+					_arrayState += 1;
```
please use explicit name of the state instead of incrementing it

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -99,6 +101,48 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
 
+J9Class *
+GC_ClassLoaderClassesIterator::nextArrayClass()
+{
```
Should it be shortcut for the case if class is not array? This code comes to right conclusion, but after switching through the three states just to return NULL

dmitripivkine replied:
It can be done in this function or at the caller side

theresa-m replied:
Do you mean check before the first state whether or not _startingClass has any arrays? Something like
```
if (NULL == _startingClass->arrayClass && NULL == _startingClass->nullRestrictedArrayClass) {
    return null;
}
```

dmitripivkine replied:
Yes, something like this whatever correct check is. Alternatively, If it is more accurate, you can add similar check on caller's side - to not call this function at all or set iterator state "done" right away instead of set initial set.

theresa-m replied:
It could be like this for the first state to avoid extra loops:
```
			case state_valueTypeArray:
				if (NULL != _startingClass->nullRestrictedArrayClass) {
					_iterateArrayClazz = _startingClass->nullRestrictedArrayClass;
					_arrayState = state_valueTypeArrayList;
				} else if (NULL != _startingClass->arrayClass) {
					_iterateArrayClazz = _startingClass->arrayClass;
					_arrayState = state_arrayList;
				} else {
					_iterateClazz = NULL;
					_arrayState = state_done;
				}
				break;
```

dmitripivkine replied:
yes, this is good

hangshao0 replied:
> Something like
> ```
> if (NULL == _startingClass->arrayClass && NULL == _startingClass->nullRestrictedArrayClass) {
>     return null;
> }
> ```
If you need to do that, please always put `_startingClass->nullRestrictedArrayClass` inside 
`if J9_IS_J9CLASS_VALUETYPE(_startingClass) {`

theresa-m replied:
Thanks @hangshao0, good idea. I added the `J9CLASS_GET_NULLRESTRICTED_ARRAY` and am using it for all accesses to nullRestrictedArrayClass except in createramclass.cpp since the ram class is not fully formed here.

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -99,6 +102,52 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
+		if (NULL != _iterateArrayClazz) {
+			break;
+		}
+	}
```
It is ok as-is, but you can reorganize this code even more. Currently you have two states for NULL restricted - `state_valueTypeArray` where you setup list walk and `state_valueTypeArrayList` where list is walked. Similarly you can have two states for ArrayList too. You do have  `state_arrayList` where list is walked, you can add `state_array` where this walk is setup. In this case  you can move `_iterateArrayClazz = _startingClass->arrayClass;...` there. The reason for this suggestion technically `state_valueTypeArrayList` or `state_valueTypeArray` should not know details related to `state_arrayList`, just know which state is next. I think it helps to avoid code duplication and make logic cleaner.
You can move this code to the new `state_array`:
```
				if (NULL != _startingClass->arrayClass) {
					_iterateArrayClazz = _startingClass->arrayClass;
					_arrayState = state_arrayList;
				} else {
					_iterateArrayClazz = NULL;
					_arrayState = state_done;
				}
```
and reuse in `state_valueTypeArrayList` too.

theresa-m replied:
Thanks for the feedback, please check the latest updates to make sure it's what you had in mind.

The first state still knows some array list details `state_valueTypeArray`, this was previously added to avoid unnecessary switches when no array lists exist.
```
				if (NULL != _startingClass->arrayClass) {
					_arrayState = state_array;
				} else {
					_arrayState = state_done;
				}
```

dmitripivkine replied:
Yes, very good, thank you! With simplification for `state_valueTypeArray` state it looks clean and easy to read.

dmitripivkine replied:
If I write this I will use `NULL !=` instead of `NULL ==` and swap halves of if. So, if condition is TRUE continue to iterate in the same state, and do state switch in the `else` part. But this is stylish thing, I am not insisting :-)

theresa-m replied:
That looks better to me to, I rearranged the logic.

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -99,6 +102,57 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
+					_arrayState = state_array;
+				} else {
+					_arrayState = state_done;
+				}
```
You can replace 
```
    else
...
    if (NULL != _startingClass->arrayClass) {
        _arrayState = state_array;
    } else {
        _arrayState = state_done;
    }
```
to
```
    else
...
    {
        _arrayState = state_array;
    }
```
I'm ok with this as a shortcut. Checks are the same, just with one extra state and one while loop iteration

theresa-m replied:
Updated.

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.hpp:
```diff
@@ -62,6 +62,16 @@ class GC_ClassLoaderClassesIterator
+		state_arrayList, /* array list walk */
+		state_done
+	};
+	ArrayClassState _arrayState;
```
Would you please add short annotation to added variables `_iterateArrayClazz`, `_startingClass` and `_arrayState` in format `/**< ... */`?

theresa-m replied:
Updated.

dmitripivkine commented:
@keithc-ca Would you please review DDR change?

dmitripivkine reviewed (approved):
GC part of the change looks good to me. Thank you!

hangshao0 commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCClassArrayClassSlotIterator.java:
```diff
@@ -47,6 +50,13 @@ protected GCClassArrayClassSlotIterator(J9ClassPointer clazz) throws CorruptData
 			slots.add(slot);
 			addresses.add(VoidPointer.cast(clazz.arrayClassEA()));
 		}
+		if (valueTypeHelper.isJ9ClassAValueType(clazz)) {
```
If the clazz is an array, this check will be false. Isn't the arrayClass->companionArray slot missed ?

hangshao0 commented on runtime/oti/j9.h:
```diff
@@ -348,14 +349,18 @@ static const struct { \
 		(J9_ARE_NO_BITS_SET((romFieldShape)->modifiers, J9AccVolatile) || (J9CLASS_UNPADDED_INSTANCE_SIZE(fieldClazz) <= sizeof(U_64))))
 #define J9_VALUETYPE_FLATTENED_SIZE(clazz) (J9CLASS_HAS_4BYTE_PREPADDING((clazz)) ? ((clazz)->totalInstanceSize - sizeof(U_32)) : (clazz)->totalInstanceSize)
+#define J9_IS_J9ARRAYCLASS_NULL_RESTRICTED(clazz) J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassArrayIsNullRestricted)
+#define J9CLASS_GET_NULLRESTRICTED_ARRAY(clazz) (J9_IS_J9CLASS_VALUETYPE(clazz) ? clazz->nullRestrictedArrayClass: NULL)
```
`clazz->nullRestrictedArrayClass` --> `(clazz)->nullRestrictedArrayClass`

keithc-ca commented on debugtools/DDR_VM/data/superset-constants.dat:
```diff
@@ -424,6 +424,7 @@ C|Illegal
 C|InlineContiguous
 S|GC_ClassArrayClassSlotIterator|GC_ClassArrayClassSlotIteratorPointer|
 C|classArrayClassSlotIterator_state_arrayClass
+C|classArrayClassSlotIterator_state_companionArray
```
Additions are not permitted to this file; see line 21:
```
# This file lists constants that have always been available.
```

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -99,6 +102,55 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
+			case state_valueTypeArray:
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+				if (NULL != J9CLASS_GET_NULLRESTRICTED_ARRAY(_startingClass)) {
+					_iterateArrayClazz = J9CLASS_GET_NULLRESTRICTED_ARRAY(_startingClass);
```
I noticed `J9CLASS_GET_NULLRESTRICTED_ARRAY()` does check for `J9_IS_J9CLASS_VALUETYPE()`, so we are doing it twice. May be is make sense to call macro only once?
```
    J9Class *nullRestrictedArray = J9CLASS_GET_NULLRESTRICTED_ARRAY(_startingClass);
    if (NULL != nullRestrictedArray) {
        _iterateArrayClazz = nullRestrictedArray;
```

theresa-m replied:
I'm getting a compile error when I try this so the declaration of `J9Class *nullRestrictedArray` will go outside the for loop. Does that sound okay?

dmitripivkine replied:
Yes, you can move this line outside while loop, it is ok too. Regarding my original suggestion I think if you put `{}` around switch case body it should resolve compile problem.
```
			case state_valueTypeArray:
            {
#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
                J9Class *nullRestrictedArray = J9CLASS_GET_NULLRESTRICTED_ARRAY(_startingClass);
                if (NULL != nullRestrictedArray) {
.....
            }
            break;
```

theresa-m replied:
Okay I'll add the brackets instead.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/AuxFieldInfo29.dat:
```diff
@@ -90,6 +90,7 @@ J9Class.instanceLeafDescription = required
 J9Class.lockOffset = required
 J9Class.module = required
 J9Class.nextClassInSegment = required
+J9Class.nullRestrictedArrayClass = required
```
This field cannot be `required` (it's only present when `defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)`). I expect we want:
```
J9Class.nullRestrictedArrayClass = J9Class*
```
The DDR code which uses this field must anticipate `NoSuchFieldException`.

theresa-m replied:
I added a new method ValueTypeHelper.areFlattenableValueTypesSupported to use around new fields so NoSuchFieldException shouldn't happen. I'm wrapping a potential NoSuchFieldException error in CorruptDataException since if that were to happen now something is pretty wrong.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCClassArrayClassSlotIterator.java:
```diff
@@ -28,11 +28,14 @@
 
 public class GCClassArrayClassSlotIterator extends GCIterator
 {
+	private ValueTypeHelper valueTypeHelper = ValueTypeHelper.getValueTypeHelper();
```
This can be static and final.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCClassArrayClassSlotIterator.java:
```diff
@@ -47,6 +50,13 @@ protected GCClassArrayClassSlotIterator(J9ClassPointer clazz) throws CorruptData
 		}
+		if (valueTypeHelper.isJ9ClassAValueType(clazz)) {
+			slot = clazz.nullRestrictedArrayClass();
+			if(slot.notNull()) {
```
Space before `(`, please.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCClassArrayClassSlotIterator.java:
```diff
@@ -47,6 +50,13 @@ protected GCClassArrayClassSlotIterator(J9ClassPointer clazz) throws CorruptData
+				addresses.add(VoidPointer.cast(clazz.nullRestrictedArrayClassEA()));
+			}
+		}
 		if(clazz.romClass().modifiers().allBitsIn(J9AccClassArray)) {
```
While you're here, can you please fix the missing space before `(` on this line as well?

keithc-ca commented on runtime/codert_vm/cnathelp.cpp:
```diff
@@ -3610,11 +3605,10 @@ fast_jitCheckCast(J9VMThread *currentThread, J9Class *castClass, j9object_t obje
-#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
+	/* In the future Valhalla checkcast needs to throw exception on
+	 * null restricted checkedType if obj is null,
+	 * see issue https://github.com/eclipse-openj9/openj9/issues/19764
```
Please correct the reference to `object` and finish each sentence with a period.

keithc-ca commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -42,6 +42,9 @@ GC_ClassLoaderClassesIterator::GC_ClassLoaderClassesIterator(MM_GCExtensionsBase
 		_mode = ANONYMOUS_CLASSES;
 	}
 	_nextClass = firstClass();
+	_iterateArrayClazz = NULL;
+	_startingClass = _nextClass;
+	_arrayState = state_valueTypeArray;
```
Please use field initializers instead of assignments where possible.

keithc-ca commented on runtime/gc_structs/ClassLoaderClassesIterator.hpp:
```diff
@@ -99,6 +109,16 @@ class GC_ClassLoaderClassesIterator
+	 * and arrayClass. nullRestrictedArrayClass can only exist as a field of a value class
+	 * and can only be an array of arity 1.
+	 * After that its list will continue in the arrayClass field.
+	 * @return the next array class, or null if finished
```
nit: `null` -> `NULL`

keithc-ca commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -764,15 +764,67 @@ JVM_IsImplicitlyConstructibleClass(JNIEnv *env, jclass cls)
+	}
+
+	if (NULL == J9CLASS_GET_NULLRESTRICTED_ARRAY(ramClass)) {
+		J9ROMArrayClass *arrayOfObjectsROMClass = (J9ROMArrayClass*)J9ROMIMAGEHEADER_FIRSTCLASS(vm->arrayROMClasses);
```
Space before `*` (in cast).

keithc-ca commented on runtime/oti/VMHelpers.hpp:
```diff
@@ -1797,6 +1798,11 @@ class VM_VMHelpers
 				}
 			}
 		}
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+		else if (J9_IS_J9ARRAYCLASS_NULL_RESTRICTED(arrayClass)) {
+			rc = FALSE;
+		}
+#endif /* if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
```
Suggest moving the closing brace after the conditional:
```c
#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
		} else if (J9_IS_J9ARRAYCLASS_NULL_RESTRICTED(arrayClass)) {
			rc = FALSE;
#endif /* if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
		}
```

keithc-ca commented on runtime/oti/j9.h:
```diff
@@ -331,7 +331,8 @@ static const struct { \
 #define J9CLASS_UNPADDED_INSTANCE_SIZE(clazz) J9_VALUETYPE_FLATTENED_SIZE(clazz)
-/* TODO replace with J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(clazz) J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassAllowsInitialDefaultValue)*/
+#define J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(clazz) J9_ARE_ALL_BITS_SET((clazz)->classFlags, J9ClassAllowsInitialDefaultValue)
+/* TODO replace with J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE */
```
Please explain (or remove) this comment.

theresa-m replied:
This was a reminder from before the q types removal work had been started https://github.com/eclipse-openj9/openj9/issues/18157. It would be removed when the J9_IS_J9CLASS_PRIMITIVE_VALUETYPE macro goes away but I don't think its really needed.

keithc-ca commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -4800,6 +4811,7 @@ typedef struct J9InternalVMFunctions {
 	void  (JNICALL *sendInit)(struct J9VMThread *vmContext, j9object_t object, struct J9Class *senderClass, UDATA lookupOptions) ;
 	void  ( *internalAcquireVMAccessNoMutex)(struct J9VMThread * vmThread) ;
 	struct J9Class*  ( *internalCreateArrayClass)(struct J9VMThread* vmThread, struct J9ROMArrayClass* romClass, struct J9Class* elementClass) ;
+	struct J9Class*  ( *internalCreateArrayClassWithOptions)(struct J9VMThread* vmThread, struct J9ROMArrayClass* romClass, struct J9Class* elementClass, UDATA options) ;
```
Space before `*`, not after
```c
	struct J9Class *(*internalCreateArrayClassWithOptions)(struct J9VMThread *vmThread, struct J9ROMArrayClass *romClass, struct J9Class *elementClass, UDATA options);
```

keithc-ca commented on runtime/oti/vm_api.h:
```diff
@@ -357,6 +357,18 @@ peekClassHashTable(J9VMThread* currentThread, J9ClassLoader* classLoader, U_8* c
 J9Class*
 internalCreateArrayClass(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Class* elementClass);
 
+/**
+* @brief
+* @param vmThread
+* @param romClass
+* @param elementClass
+* @param options
+* @return J9Class*
+*/
+J9Class*
+internalCreateArrayClassWithOptions(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Class* elementClass, UDATA options);
+
```
Please complete this documentation.
The asterisk of the second and subsequent lines of a multi-line comment should line up with the one in the first line.
Space before, not after `*`.
Blank line 370 is not needed.

keithc-ca commented on runtime/vm/classsupport.c:
```diff
@@ -196,22 +196,27 @@ findPrimitiveArrayClass(J9JavaVM* vm, jchar sigChar)
 #if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
-	/* When creating an array of valuetype elements, the array elements are initialized to the defaultValue of the
-	 * element type. As a result the element type must be fully initialized (if its a valuetype) before creating an
+	/* When creating an array of implicitly constructible valuetype elements, the array elements are initialized to
+	 * the defaultValue of the element type. As a result the element type must be fully initialized before creating an
 	 * instance of the array. Element class init must be done before the arrayClass is created so that in the case
 	 * of an init failure the arrayClass is not temporarily exposed.
```
Please fold these overly long lines.

keithc-ca commented on runtime/vm/createramclass.cpp:
```diff
@@ -2388,7 +2400,21 @@ internalCreateRAMClassDone(J9VMThread *vmThread, J9ClassLoader *classLoader, J9C
+				if ((NULL != elementClass->nullRestrictedArrayClass) && (NULL != elementClass->arrayClass)) {
+					((J9ArrayClass *)elementClass->arrayClass)->companionArray = elementClass->nullRestrictedArrayClass;
+					((J9ArrayClass *)elementClass->nullRestrictedArrayClass)->companionArray = elementClass->arrayClass;
+
```
Please remove this blank line.

theresa-m commented:
Thanks @keithc-ca @dmitripivkine @hangshao0 , I just pushed an update addressing all the comments starting after https://github.com/eclipse-openj9/openj9/pull/19911#pullrequestreview-2251205770

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -33,15 +33,18 @@
 GC_ClassLoaderClassesIterator::GC_ClassLoaderClassesIterator(MM_GCExtensionsBase *extensions, J9ClassLoader *classLoader) :
 	_javaVM((J9JavaVM* )extensions->getOmrVM()->_language_vm)
 	,_classLoader(classLoader)
+	,_nextClass(firstClass())
```
Please don't move `_nextClass`, we should not call `firstClass()` method here, as well as `_startingClass` initialization as well (because it depends at `_nextClass` initialization obviously).

theresa-m replied:
Updated.

dmitripivkine commented:
@keithc-ca Your comments have been addressed. Do you have more suggestions?

dmitripivkine commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -99,6 +102,58 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
+GC_ClassLoaderClassesIterator::nextArrayClass()
+{
+	while (_arrayState != state_done) {
+		switch(_arrayState) {
```
Correct. Please change indentation of the code inside switch.

dmitripivkine replied:
Also I believe `{}` inside `state_valueTypeArray` state should be indented to the `case` line.

theresa-m replied:
> Also I believe `{}` inside `state_valueTypeArray` state should be indented to the `case` line.

What about the break that comes after?

dmitripivkine replied:
break should be indented right once from case line

keithc-ca replied:
Both patterns appear in the code, but my preference is to put the `break` after the braces.
```c
		switch (_arrayState) {
		case state_valueTypeArray:
			{
				// ...
			}
			break;
		case state_valueTypeArrayList:
			// ...
```

keithc-ca commented:
Please also fix the commit message: lines in the body should be no more than 72 characters long (see https://github.com/eclipse/omr/blob/master/CONTRIBUTING.md#commit-guidelines).

theresa-m commented on runtime/gc_structs/ClassLoaderClassesIterator.hpp:
```diff
@@ -62,6 +62,16 @@ class GC_ClassLoaderClassesIterator
+	J9Class *_iterateArrayClazz; /**< current array walk class */
+	J9Class *_startingClass; /**< the most recent table or system class which is the starting point for each array walk */
+	enum ArrayClassState {
+		state_valueTypeArray = 0, /**< setup value type array list walk */
+		state_valueTypeArrayList, /**< value type array list walk */
+		state_array, /**< setup array list walk */
+		state_arrayList, /**< array list walk */
+		state_done
```
This was based on the format of the state names in ClassArrayClassSlotIterator.hpp. I don't mind changing to all caps though, is there a preference for gc @dmitripivkine ?

keithc-ca replied:
This is inconsistent. As is, we will have one class with nested enums following different patterns. There's no need for that. Using block capitals (like the existing enum) signals to readers that the name refers to a constant.

theresa-m replied:
I changed the names to all caps.

theresa-m commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCClassArrayClassSlotIterator.java:
```diff
@@ -43,19 +46,30 @@ protected GCClassArrayClassSlotIterator(J9ClassPointer clazz) throws CorruptData
 			J9ArrayClassPointer arrayClass = J9ArrayClassPointer.cast(clazz);
+			if (valueTypeHelper.areFlattenableValueTypesSupported()) {
+				try {
+					slot = arrayClass.companionArray();
```
I believe this was fixed already, can you double check your commit it the latest?

keithc-ca commented on runtime/gc_structs/ClassLoaderClassesIterator.cpp:
```diff
@@ -99,6 +102,58 @@ GC_ClassLoaderClassesIterator::switchToSystemMode()
+		switch(_arrayState) {
+		case STATE_VALUETYPEARRAY:
+			{
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			J9Class *nullRestrictedArray = J9CLASS_GET_NULLRESTRICTED_ARRAY(_startingClass);
+			if (NULL != nullRestrictedArray) {
+				_iterateArrayClazz = nullRestrictedArray;
+				_arrayState = STATE_VALUETYPEARRAYLIST;
+			} else
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
+			{
+				_arrayState = STATE_ARRAY;
+			}
```
This should be indented one more level.

keithc-ca commented on runtime/oti/vm_api.h:
```diff
@@ -357,6 +357,19 @@ peekClassHashTable(J9VMThread* currentThread, J9ClassLoader* classLoader, U_8* c
+/**
+ * @brief Create a new J9Class to represent an array of elementClass.
+ *
+ * @param vmThread current vm thread
+ * @param romClass the rom class associated with new RAM class. For arrays
+ * this is always the object array ROM class "[L".
+ * @param elementClass element or base class or array to be created
```
See https://github.com/eclipse-openj9/openj9/pull/19911#discussion_r1731595654.

theresa-m replied:
Looks like I missed a ROM capitalization. Was there anything else?

keithc-ca replied:
Have you addressed all other comments in my [latest review](https://github.com/eclipse-openj9/openj9/pull/19911#pullrequestreview-2261662850)?

theresa-m replied:
I meant for this method descriptor specifically, but yes I just finished with all the latest review comments.

keithc-ca commented on runtime/oti/vm_api.h:
```diff
@@ -357,6 +357,19 @@ peekClassHashTable(J9VMThread* currentThread, J9ClassLoader* classLoader, U_8* c
+ * @return J9Class* J9Class of an elementClass array
+ */
+J9Class *
+internalCreateArrayClassWithOptions(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Class* elementClass, UDATA options);
```
Space before, not after each `*`.

keithc-ca commented on runtime/vm/classsupport.c:
```diff
@@ -196,22 +196,29 @@ findPrimitiveArrayClass(J9JavaVM* vm, jchar sigChar)
  */
 J9Class* 
 internalCreateArrayClass(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Class* elementClass)
+{
+	return internalCreateArrayClassWithOptions(vmThread, romClass, elementClass, 0);
+}
+
+J9Class*
+internalCreateArrayClassWithOptions(J9VMThread *vmThread, J9ROMArrayClass *romClass, J9Class *elementClass, UDATA options)
```
Space before, not after each `*`.

keithc-ca commented on runtime/vm/classsupport.c:
```diff
@@ -196,22 +196,29 @@ findPrimitiveArrayClass(J9JavaVM* vm, jchar sigChar)
 	j9object_t heapClass = J9VM_J9CLASS_TO_HEAPCLASS(elementClass);
 	j9object_t protectionDomain = NULL;
-	J9ROMClass* arrayRomClass = (J9ROMClass*) romClass;
+	J9ROMClass *arrayRomClass = (J9ROMClass*) romClass;
```
Space before each `*`, no space after `)` (compare with cast on line 223).

keithc-ca commented:
The commit message should also hyphenate "null-restricted".

keithc-ca commented:
I also noticed a typo in the commit message: "one doesn't exists" should be "one doesn't exist".

keithc-ca commented on runtime/oti/vm_api.h:
```diff
@@ -357,6 +357,19 @@ peekClassHashTable(J9VMThread* currentThread, J9ClassLoader* classLoader, U_8* c
+/**
+ * @brief Create a new J9Class to represent an array of elementClass.
+ *
+ * @param vmThread current vm thread
+ * @param romClass the ROM class associated with new RAM class. For arrays
+ * this is always the object array ROM class "[L".
```
* "vm" should be capitalized, see line 376

* `@param` clauses don't normally end with a period, suggest parentheses:
```
 * @param romClass the ROM class associated with new RAM class (for arrays
 * this is always the object array ROM class "[L")
```

Please correct "null restricted" to "null-restricted" throughout this file and others changed by this pull request.

keithc-ca commented on runtime/vm/classsupport.c:
```diff
@@ -196,22 +196,29 @@ findPrimitiveArrayClass(J9JavaVM* vm, jchar sigChar)
  */
 J9Class* 
 internalCreateArrayClass(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Class* elementClass)
+{
```
Space before, not after, each `*`.

keithc-ca reviewed (changes requested):
In the files changed by this pull request, there are still several occurrences of "null restricted" that should be "null-restricted".

Please also address the line-end check failure.

keithc-ca commented:
> Which others are you seeing

* `vm_api.h` - lines 2715, 2719, 2823, etc.
* `BytecodeInterpreter.hpp` - line 8585 and missing comma on line 8584
* `ValueTypeHelpers.hpp` - lines 193, 197
* `ValueTypeTest.java` - lines 196, 1804, 2262

theresa-m commented:
> > Which others are you seeing
> 
> * `vm_api.h` - lines 2715, 2719, 2823, etc.
> * `BytecodeInterpreter.hpp` - line 8585 and missing comma on line 8584
> * `ValueTypeHelpers.hpp` - lines 193, 197
> * `ValueTypeTest.java` - lines 196, 1804, 2262

Do you mind if I do this as a separate pull requests? This is out of scope of the current change since none of those code sections are being modified.

theresa-m commented:
Okay. I have fixed the remaining cases in https://github.com/eclipse-openj9/openj9/pull/20085

keithc-ca commented:
Jenkins test sanity aarch64_linux_vt_standard,s390x_linux

hangshao0 commented:
Jenkins test sanity,extended zlinuxval jdknext

theresa-m commented:
I'm looking at the test failures.

theresa-m commented:
1. For the sanity failures many of them seem to already exist. I ran sanity tests for a value types build on the master branch and similar failures occurred. https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/23714/
I'm not sure where that started since the valhalla nightly builds don't run full sanity tests, I will need to do more testing.

2. I pushed a second commit to fix the extended failures. The tests are currently written to use arrayClass so to keep them passing both arrayClass and nullRestrictedArrayClass can have the J9ClassIsFlattened set for now. The purpose of this pull request was to add support for null-restricted arrays without flattening support. I have a draft pull request https://github.com/eclipse-openj9/openj9/pull/19995 that will address flattening where I update the value type command line tests and will stop setting J9ClassIsFlattened for arrayClass.


failing pr build links for reference:
https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/5985/
https://openj9-jenkins.osuosl.org/job/Test_openjdknext_j9_sanity.functional_aarch64_linux_vt_standard_Personal/8/
https://openj9-jenkins.osuosl.org/job/Test_openjdknext_j9_sanity.functional_s390x_linux_Personal/62/
https://openj9-jenkins.osuosl.org/job/Test_openjdkValhalla_j9_extended.functional_s390x_linux_valhalla_Personal/27/
https://openj9-jenkins.osuosl.org/job/Test_openjdkValhalla_j9_sanity.functional_s390x_linux_valhalla_Personal/29/

hangshao0 commented:
Jenkins test sanity,extended zlinuxval jdknext

hangshao0 commented:
Jenkins compile amac jdknext

theresa-m commented:
The failures in ValueTypeTestsJIT are related to https://github.com/eclipse-openj9/openj9/pull/19911/commits/42393ffc973df02038ba16a9f0a80d559994f18d and array flattening. I have disabled them for now since there is more work to be done to flatten arrays in https://github.com/eclipse-openj9/openj9/pull/19995

hangshao0 commented:
Jenkins test sanity,extended zlinuxval jdknext

pshipton commented:
I believe this is causing hangs in testing. I'll revert it while we investigate.

pshipton commented:
Reverted via https://github.com/eclipse-openj9/openj9/pull/20109
